### PR TITLE
Add Sys.which and Sys.isexecutable

### PR DESCRIPTION
**Pull request by ararslan**

### CODE REVIEW

**Comment by martinholters** [comment]
Should wait whether https://github.com/JuliaLang/julia/issues/27284 requires adjustments here.

File: src/Compat.jl [hunk 1]
```diff
@@ -461,6 +461,45 @@ module Sys
...
+                append!(dirs, map(abspath, split(get(ENV, "PATH", ""), pathsep)))
+                iswindows() && pushfirst!(dirs, pwd())
+            else
+                push!(dirs, abspath(dir))
+            end
+            for d in dirs, p in progs
+                path = joinpath(d, p)
+                isexecutable(path) && return realpath(path)
+            end
+            error("$program not found")
```
Comment (martinholters) [inline src/Compat.jl]: Should be `return nothing` now?

File: README.md [hunk 2]
```diff
@@ -295,6 +295,8 @@ Currently, the `@compat` macro supports the following syntaxes:
 
 * `squeeze` with `dims` as keyword argument ([#26660]).
 
+* `Compat.Sys.which` and `Compat.Sys.isexecutable` ([#26559]).
```
Comment (martinholters) [inline README.md]: And `#27284`.

**Comment by ararslan** [comment]
I added logic to handle the case between the two PRs.

**Comment by ararslan** [comment]
CI failures on 0.7 will likely be fixed by #563.

**Comment by martinholters** [comment]
I'll close/open to re-trigger CI now that #563 is in.

**Comment by martinholters** [comment]
On 0.6/windows:
```
Error During Test
  Test threw an exception of type UndefVarError
  Expression: Compat.Sys.which("julia") == realpath(jl)
  UndefVarError: pushfirst! not defined
  Stacktrace:
   [1] which(::String) at C:\Users\appveyor\.julia\v0.6\Compat\src\Compat.jl:483
```
The line number seems to be off, `pushfirst!` actually occurs in line 481.

File: src/Compat.jl [hunk 3]
```diff
@@ -453,6 +453,54 @@ module Sys
...
+                end
+            else
+                push!(progs, base)
+            end
+            dirs = String[]
+            dir = dirname(program)
+            if isempty(dir)
+                pathsep = iswindows() ? ';' : ':'
+                append!(dirs, map(abspath, split(get(ENV, "PATH", ""), pathsep)))
+                iswindows() && pushfirst!(dirs, pwd())
```
Comment (martinholters) [inline src/Compat.jl]: This is in a submodule, so `Compat.pushfirst!` isn't bound to `pushfirst!` here.

Comment (ararslan) [reply src/Compat.jl]: I think I've fixed it now.

**Pull request merged**